lib/Sema/ConstraintSystem.h

```diff
@@ -3486,6 +3486,50 @@ class OverloadSetCounter : public ASTWalker {
     return { true, expr };
   }
 };
+
+
+/// \brief Matches array of function parameters to candidate inputs,
+/// which can be anything suitable (e.g., parameters, arguments).
+///
+/// It claims inputs sequentially and tries to pair between an input
+/// and the next appropriate parameter. The detailed matching behavior
+/// of each pair is specified by a custom function (i.e., pairMatcher).
+/// It considers variadic and defaulted arguments when forming proper
+/// input-parameter pairs; however, other information like label and
+/// type information is not directly used here. It can be implemented
+/// in the custom function when necessary.
+class InputMatcher {
+  size_t NumSkippedParameters;
+  const llvm::SmallBitVector DefaultValueMap;
+  const ArrayRef<AnyFunctionType::Param> Params;
+
+public:
+  enum Result {
+    /// The specified inputs are successfully matched.
+    IM_Succeeded,
+    /// There are input(s) left unclaimed while all parameters are matched.
+    IM_HasUnclaimedInput,
+    /// There are parateter(s) left unmatched while all inputs are claimed.
+    IM_HasUnmatchedParam,
+    /// Custom pair matcher function failure.
+    IM_CustomPairMatcherFailed,
+  };
+
+  InputMatcher(const ArrayRef<AnyFunctionType::Param> params,
+      const llvm::SmallBitVector &defaultValueMap);
```

Comment:
Ok, here is an idea - instead of accepting a list of params here, let's give `InputMatcher` `AbstractFunctionDecl` where `ParamDecl`s could be extracted via `getParameterList` (just like what `paramisIUO` does today), and let's make it so matcher callback uses `ParamDecl` everywhere, these decls should have all the information needed -label, type and flags. Let's try that, if it doesn't work out we'll go with current approach, but if it does that would be a very nice cleanup!

Reply:
I tried and it seems not easy to achieve this... 

The list of params passed to `InputMatcher` in CSRanking.cpp is from [funcTy2 / openType2](https://github.com/apple/swift/pull/17093/files#diff-a72024f35020a11c7f1e09bcd35b22bdL608), which is not directly computed from `decl2` and the computation is non-trivial (involving `cs` intermediately). If we'd like to compute such list of params locally in `InputMatcher`, the code for such computation would be bulky and some extra stuffs (e.g., `cs`) should also be passed to `InputMatcher`.

In addition, if the list of params is not prepared by the caller, we would have to pass additional information  to `InputMatcher`(like [ignoreLastPair](https://github.com/dingobye/swift/blob/93dfa5b89658ac199b002e7c8b5507617fc1f748/lib/Sema/ConstraintSystem.h#L3528) in a previous version of this patch), which you dislike as in your previous comment on 19th June.

It seems we do not have much choice here. Shall we just use indices?

Reply:
Alright, fair enough!